tianshou/data/collector.py

```diff
@@ -320,3 +369,24 @@ def sample(self, batch_size: int) -> Batch:
         batch_data, indice = self.buffer.sample(batch_size)
         batch_data = self.process_fn(batch_data, self.buffer, indice)
         return batch_data
+
+
+def _batch_set_item(source: Batch, indices: np.ndarray,
+                    target: Batch, size: int):
+    # for any key chain k, there are three cases
+    # 1. source[k] is non-reserved, but target[k] does not exist or is reserved
+    # 2. source[k] does not exist or is reserved, but target[k] is non-reserved
+    # 3. both source[k] and target[k] is non-reserved
+    for k, v in target.items():
+        if not isinstance(v, Batch) or not v.is_empty():
+            # target[k] is non-reserved
+            vs = source.get(k, Batch())
+            if isinstance(vs, Batch) and vs.is_empty():
+                # case 2
+                # use __dict__ to avoid many type checks
+                source.__dict__[k] = _create_value(v[0], size)
+        else:
+            # target[k] is reserved
+            # case 1
+            continue
+        source.__dict__[k][indices] = v
```

Comment:
try/except should be more efficient, since the other cases seldom occur.
```
    for k, v in target.items():
        try:
            source.__dict__[k][indices] = v
        except IndexError:
            if not isinstance(v, Batch) or not v.is_empty():
                # target[k] is non-reserved
                vs = source.get(k, Batch())
                if isinstance(vs, Batch) and vs.is_empty():
                    # case 2
                    # use __dict__ to avoid many type checks
                    source.__dict__[k] = _create_value(v[0], size)
            source.__dict__[k][indices] = v
```

But as you want.

Reply:
I feel uncomfortable with explicitly invoking an exception and then catching it so I write it this way 😄

Reply:
I few months ago I would have said exactly the same, but it turns out try/except is way more efficient in practice if the exception is very seldom raised, as it should be the case here. You should have a look at [this post](https://stackoverflow.com/a/7604717/4820605) for instance. But it is up to you !

Reply:
> it turns out try/except is way more efficient in practice if the exception is very seldom raised

That's an interesting point. The post is interesting, too. If someday I meet efficiency bottleneck then I'll consider something like this, but if not, I prefer the straightforward way to make it clear for future maintenance.